### PR TITLE
feat: add *Nushell* support

### PR DESCRIPTION
**Pull request by stevenxxiu**
See [NuShell support · Issue #687 · ellie/atuin](https://github.com/ellie/atuin/issues/687).

### CODE REVIEW

**Comment by ellie** [comment]
I'll give this a proper test and review tomorrow! Thank you 🙏

File: atuin-client/src/import/nu_histdb.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+impl From<HistDbEntry> for History {
+    fn from(histdb_item: HistDbEntry) -> Self {
+        let ts_secs = histdb_item.start_timestamp / 1000;
+        let ts_ns = (histdb_item.start_timestamp % 1000) * 1_000_000;
+        History::new(
+            DateTime::from_utc(NaiveDateTime::from_timestamp(ts_secs, ts_ns as u32), Utc),
+            String::from_utf8(histdb_item.command_line)
+                .unwrap_or_else(|_e| String::from(""))
```
Comment (ellie) [inline atuin-client/src/import/nu_histdb.rs:36]: I'd rather skim adding the entry vs add a blank one, might as well avoid noise in the db!

**Review by ellie — COMMENTED** [review]
Couple of comments - I'm going to install nu now as well, just to get a bit of history

File: README.md [hunk 2]
```diff
@@ -274,6 +274,16 @@ Install `atuin` shell plugin in zsh, bash, or fish with [Fig](https://fig.io) in
 
 <a href="https://fig.io/plugins/other/atuin" target="_blank"><img src="https://fig.io/badges/install-with-fig.svg" /></a>
 
+### Nushell
+
+Add to `config.nu`:
+
+```
+mkdir ~/.local/share/atuin/
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply README.md:282]: I believe that's what "Run in Nushell" is trying to suggest, it's the install setup, not yet in the config.nu

The "add to config.nu" section below has only the `source init.nu` line

[the remark this answers is not in the document]
Comment (conradludgate) [reply README.md:282]: Oh. Github mobile is trash. This comment showed even after Steven updated the code 😂. It looks resolved now :)

**Review by conradludgate — COMMENTED** [review]
Code looks good. I'll have a test soon and see how it feels. Thank you so much ❤️

**Comment by stevenxxiu** [comment]
Rebased due to [Add history deletion by ellie · Pull Request #791 · ellie/atuin](https://github.com/ellie/atuin/pull/791).

**Comment by ellie** [comment]
Testing this on my machine atm - it looks like perhaps new history items aren't being stored OK? I'm running commands and nothing is stored, though I am occasionally seeing blank entries. Perhaps the precmd is running OK but post command is not? I'm running `nu 0.77.1` so I think that should be ok

And also, the up arrow binding is not bound by default - would be awesome to get the behaviour matching our defaults elsewhere!

Otherwise everything is super smooth, thank you!

**Comment by stevenxxiu** [comment]
Oh you need 0.77.2 as don't think 0.77.1 includes my hook PR.

Sure I can include the up arrow keybind too. Good idea.

**Comment by stevenxxiu** [comment]
I don't actually think getting the edit buffer is currently possible and pass it to *Atuin*, due to the *Nushell* issues:

- [The `complete` command doesn't work with interactive commands · Issue #8539 · nushell/nushell](https://github.com/nushell/nushell/issues/8539)
- [Redirect `stderr` · Issue #8277 · nushell/nushell](https://github.com/nushell/nushell/issues/8277)

**Comment by stevenxxiu** [comment]
> Actually let me see if I can hack around this a bit. Maybe I can use *Sh* to swap stdout and stderr.

This is doable, but an even bigger problem is that I'm not sure how and if it's possible to get the current input buffer's contents, in a `executehostcommand`. I'm asking *Nushell* people about this.

**Comment by stevenxxiu** [comment]
Requires my PR https://github.com/nushell/nushell/pull/8560.

File: src/command/init.rs [hunk 3]
```diff
@@ -90,11 +92,43 @@ bind -M insert \e\[A _atuin_bind_up";
...
+    $env.config | upsert keybindings (
+        $env.config.keybindings
+        | append {
+            name: atuin
+            modifier: control
+            keycode: char_r
+            mode: emacs
+            event: { send: executehostcommand cmd: (_atuin_search_cmd) }
+        }
+        | append {
```
Comment (stevenxxiu) [inline src/command/init.rs:111]: `BIND_UP_ARROW` separately.

File: src/shell/atuin.nu [hunk 4]
```diff
@@ -1,8 +1,14 @@
 # Source this in your ~/.config/nushell/config.nu
 let-env ATUIN_SESSION = (atuin uuid)
 
+# Magic token to make sure we don't record commands run by keybindings
+let ATUIN_KEYBINDING_TOKEN = $"# (random uuid)\n"
+
 let _atuin_pre_execution = {||
-    let-env ATUIN_HISTORY_ID = (atuin history start -- (commandline))
+    let cmd = (commandline)
+    if $ATUIN_KEYBINDING_TOKEN not-in $cmd {
```
Comment (stevenxxiu) [inline src/shell/atuin.nu:9]: Use `str starts-with` and not `not-in`.

File: src/shell/atuin.nu [hunk 5]
```diff
@@ -17,10 +23,13 @@ let _atuin_pre_prompt = {||
...
-        $flags,
-        ` -i -- "$0" 3>&1 1>&2 2>&3' (commandline) | str substring ',-1')`,
-    ] | flatten | str join ''
+        $ATUIN_KEYBINDING_TOKEN,
+        ([
+            `commandline (sh -c 'RUST_LOG=error atuin search `,
+            $flags,
+            ` -i -- "$0" 3>&1 1>&2 2>&3' (commandline) | str substring ',-1')`,
+        ] | flatten | str join ''),
+    ] | str join '; '
```
Comment (stevenxxiu) [inline src/shell/atuin.nu:32]: Use a newline instead of `str join '; '`.

**Comment by stevenxxiu** [comment]
@ellie Yup it should be. I'm guessing *Nushell* `master` is the easiest. Personally I'm just running v0.77.2 with a few patches including this one.

**Comment by fdncred** [comment]
Hi! I'm one of the nushell maintainers. I'm excited to use nushell with atuin. However, it breaks my heart a little to see it not be cross platform because of this one part.
```
def _atuin_search_cmd [...flags: string] {
    [
        $ATUIN_KEYBINDING_TOKEN,
        ([
            `commandline (sh -c 'RUST_LOG=error atuin search `,
            $flags,
            ` -i -- "$0" 3>&1 1>&2 2>&3' (commandline))`,
        ] | flatten | str join ''),
    ] | str join "\n"
}
```
Can you help me understand the technical reason why this needs to happen? I'm not a regular bash user but I think this is swapping stderr and stdout, but tbh, I'm not positive. Is there something that nushell could do to help make this work better?

**Comment by stevenxxiu** [comment]
@fdncred Yes this is just swapping stderr and stdout. I believe the interface is on stdout, and the command that's selected goes to stderr. This gets swapped so stdout contains the command that's selected.

From my point of view these two issues need to be addressed:

- [The `complete` command doesn't work with interactive commands · Issue #8539 · nushell/nushell](https://github.com/nushell/nushell/issues/8539)
- [Redirect `stderr` · Issue #8277 · nushell/nushell](https://github.com/nushell/nushell/issues/8277)

**Comment by conradludgate** [comment]
I'll test if we can switch them in other shells to make things easier

File: README.md [hunk 6]
```diff
@@ -274,6 +274,21 @@ Install `atuin` shell plugin in zsh, bash, or fish with [Fig](https://fig.io) in
 
 <a href="https://fig.io/plugins/other/atuin" target="_blank"><img src="https://fig.io/badges/install-with-fig.svg" /></a>
 
+### Nushell
+
+Run in *Nushell*:
+
+```
+mkdir ~/.local/share/atuin/
+atuin init nushell | save ~/.local/share/atuin/init.nu
```
Comment (ellie) [inline README.md:283]:
```suggestion
atuin init nu | save ~/.local/share/atuin/init.nu
```

Comment (stevenxxiu) [reply README.md]: Thanks! Good pickup.

**Review by ellie — APPROVED** [review]
LGTM! Thank you so much for this!!! 🙏 Really appreciate the work you have done on both sides.

I'm not sure how well this will work on Windows (and on which terminals/setups it will work), but we don't claim solid Windows support either way - especially given how fresh thast support is.

**Comment by happysalada** [comment]
thanks a lot for this, I've just tested this and ran into a small error
https://github.com/ellie/atuin/issues/841
happy to provide more feedback and/or test anything.

**Pull request merged**